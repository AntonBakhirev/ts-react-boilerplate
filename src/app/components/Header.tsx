import * as React from "react";
import {FormattedMessage} from "react-intl";
import {style} from "typestyle";
import {Link} from "./Link";

const styles = {
  nav: style({
    $nest: {
      ul: {
        $nest: {
          li: {
            display: "inline",
            padding: "5px"
          }
        },
        listStyleType: "none",
        padding: 0
      }
    }
  })
};

const Header = () => (
  <nav className={styles.nav}>
    <ul>
      <li><Link name="home"><FormattedMessage id="home" defaultMessage="Home" /></Link></li>
      <li><Link name="about"><FormattedMessage id="about" defaultMessage="About Us" /></Link></li>
      <li><Link name="counter"><FormattedMessage id="counter" defaultMessage="Counter" /></Link></li>
      <li><Link name="stars"><FormattedMessage id="stars" defaultMessage="Stars" /></Link></li>
      <li><Link name="register"><FormattedMessage id="register" defaultMessage="Register" /></Link></li>
    </ul>
  </nav>
);

export {Header};
