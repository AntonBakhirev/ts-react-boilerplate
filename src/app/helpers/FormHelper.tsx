import * as React from "react";
import {FormattedMessage} from "react-intl";
import {connect} from "react-redux";
import {Field, GenericField, WrappedFieldInputProps, WrappedFieldMetaProps} from "redux-form";
import {style} from "typestyle";
import {IStore} from "../redux/IStore";

interface IFlexWrappedFieldProps {
  input?: Partial<WrappedFieldInputProps>;
  meta?: Partial<WrappedFieldMetaProps>;
}

// While waiting for this pull request https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17693,
// we overrides meta as any for now
interface IRenderFieldProps extends IFlexWrappedFieldProps {
  defaultMessage: string;
  languageId: string;
  type?: string;
}

export const FieldExtra = Field as new () => GenericField<IRenderFieldProps>;

export const required = (id, defaultMessage) => (value) => (value ? undefined : {id, defaultMessage});

export const maxLength = (id, defaultMessage) => (max) => (value) => value && value.length > max ? {id, defaultMessage, values: {max}} : undefined;

export const minLength = (id, defaultMessage) => (min) => (value) => value && value.length < min ? {id, defaultMessage, values: {min}} : undefined;

export const numberType = (id, defaultMessage) => (value) => value && isNaN(Number(value)) ? {id, defaultMessage} : undefined;

export const minValue = (id, defaultMessage) => (min) => (value) => value < min ? {id, defaultMessage, values: {min}} : undefined;

export const matchedPwd = (id, defaultMessage) => (value, otherValues) => value !== otherValues.password ? {id, defaultMessage} : undefined;

export const email = (id, defaultMessage) => (value) => value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ? {id, defaultMessage} : undefined;

export const tooOld = (id, defaultMessage) => (value) => value && value > 65 ? {id, defaultMessage} : undefined;

export const aol = (id, defaultMessage) => (value) => value && /.+@aol\.com/.test(value) ? {id, defaultMessage} : undefined;

/*tslint:disable:object-literal-sort-keys*/
const styles = {
  row: style({
    margin: "15px 0"
  }),
  inputItem: style({
    padding: 5,
    border: "1px solid #ccc"
  }),
  message: style({
    marginTop: 5
  }),
  error: style({
    color: "red"
  }),
  warning: style({
    color: "darkorange"
  })
};
/*tslint:enable:object-literal-sort-keys*/

/*tslint:disable:jsx-no-multiline-js*/
class CustomField extends React.Component<IStateToProps & IRenderFieldProps, null> {
  public render(): JSX.Element {
    const {defaultMessage, input, languageId, type, meta: {active, touched, error, warning}} = this.props;
    return (
      <div className={styles.row}>
        <FormattedMessage id={languageId} defaultMessage={defaultMessage} />
        <div>
          <input {...input} placeholder={this.props.languageData[languageId] || defaultMessage} type={type} className={styles.inputItem} />
          {
            (active || touched) &&
            (
              (
                error &&
                <div className={styles.message}>
                <span className={styles.error}>
                  <FormattedMessage id={error.id} defaultMessage={error.defaultMessage} values={error.values} />
                </span>
                </div>
              )
              ||
              (
                warning &&
                <div className={styles.message}>
                <span className={styles.warning}>
                  <FormattedMessage id={warning.id} defaultMessage={warning.defaultMessage} values={warning.values} />
                </span>
                </div>
              )
            )
          }
        </div>
      </div>
    );
  }
}

interface IStateToProps {
  languageData: any;
}

const mapStateToProps = (state: Partial<IStore>) => ({
  languageData: state.language.payload.languageData
});

const ConnectedCustomField = connect<IStateToProps, null, IRenderFieldProps>(mapStateToProps, null)(CustomField);

export {CustomField as UnconnectedCustomField, ConnectedCustomField as CustomField, mapStateToProps};
