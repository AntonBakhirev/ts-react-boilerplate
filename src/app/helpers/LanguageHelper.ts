import * as fs from "fs";

export class LanguageHelper {
  private preferedLang: string;
  constructor(private requestLanguage: string) {
    this.preferedLang = this.requestLanguage.split(",")[0];
  }

  public static getDefaultLanguage(): string {
    return "en-GB";
  }

  public static isSupported(language: string): boolean {
    return LanguageHelper.getSupportedLanguagesArray().indexOf(language) !== -1;
  }

  private static getLanguageFileLocation(language: string): string {
    return `${__dirname}/../translations/${language.toLowerCase()}.json`;
  }

  private static getSupportedLanguagesArray(): string[] {
    return ["en-GB", "en-US", "de"];
  }

  private static getLanguageData(language: string): object {
    return JSON.parse(fs.readFileSync(LanguageHelper.getLanguageFileLocation(language)).toString());
  }

  public getRequestLanguageData(): any {
    return LanguageHelper.getLanguageData(LanguageHelper.isSupported(this.preferedLang) ? this.preferedLang : "en-GB");
  }

  public getRequestLang(): string {
    return this.requestLanguage;
  }

  public getPreferredLanguage(): string {
    return this.preferedLang;
  }
}
