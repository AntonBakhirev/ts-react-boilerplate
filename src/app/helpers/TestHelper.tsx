/** React Specific */
import {mount, ReactWrapper} from "enzyme";
import * as React from "react";
import {IntlProvider} from "react-intl";
import {Provider} from "react-redux";
import {createStore} from "redux";
import rootReducer from "../redux/rootReducer";

/** Redux Mock Store Configuration */
import {IStore} from "../redux/IStore";

/** Render Component */
function renderComponent(ComponentClass: React.ComponentClass<any> | React.SFC<any> | React.ClassType<any, any, any> | string, state?: object, props?: object): ReactWrapper<IStore, any> {
  const store = createStore(rootReducer, state);

  return mount(
    <IntlProvider locale="en">
      <Provider store={store}>
        <ComponentClass {...props} />
      </Provider>
    </IntlProvider>
  );
}

export {renderComponent};
