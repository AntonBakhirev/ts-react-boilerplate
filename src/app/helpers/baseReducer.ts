import IBaseAction from "../models/IBaseAction";

/**
 * @param name todo: change to a better name
 * @param state
 * @param action
 * @returns {any}
 */
export default function baseReducer<TState, TAction>(name: string, state: TState, action: IBaseAction & TAction): TState {
  switch (action.type) {
    case name + "_PENDING":
      return Object.assign({}, state, {
        isFetching: true
      });

    case name + "_FULFILLED":
      return Object.assign({}, state, {
        isFetching: false,
        payload: action.payload
      });

    case name + "_REJECTED":
      return Object.assign({}, state, {
        error: true,
        isFetching: false,
        payload: action.payload
      });

    default:
      return state;
  }
}