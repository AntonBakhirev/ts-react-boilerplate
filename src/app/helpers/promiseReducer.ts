import {IAction, IState} from "../redux/modules/baseModule";

export interface IPromiseActions {
  FULFILLED: string;
  PENDING: string;
  REJECTED: string;
}

export const promiseAction = (baseAction: string): IPromiseActions => {
  return {
    FULFILLED: baseAction + "_FULFILLED",
    PENDING: baseAction + "_PENDING",
    REJECTED: baseAction + "_REJECTED"
  };
};

// tslint:disable:max-line-length
export default function promiseReducer<P, M = {}>(baseAction: string, state: IState<P, M>, action: IAction<P, M>): IState<P, M> {
  switch (action.type) {
    case promiseAction(baseAction).PENDING:
      return {
        ...state,
        isFetching: true
      };

    case promiseAction(baseAction).FULFILLED:
      return {
        ...state,
        isFetching: false,
        payload: action.payload
      };
    case promiseAction(baseAction).REJECTED:
      return {
        ...state,
        error: true,
        isFetching: false,
        message: action.message
      };
    default:
      return state;
  }
}
