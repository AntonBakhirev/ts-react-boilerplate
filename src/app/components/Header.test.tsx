import {TestHelper} from "../helpers/TestHelper";
import {Header, Styles} from "./Header";

describe("<Header />", () => {

  const renderer = new TestHelper();
  const component = renderer.withTranslation({locale: "en-GB", languageData: {}}).mount(Header);

  it("Renders with correct style", () => {
    expect(component.find("nav")).toHaveClassName(Styles.nav);
  });

});
