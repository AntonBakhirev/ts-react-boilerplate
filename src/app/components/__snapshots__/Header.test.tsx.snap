// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> matches snapshot 1`] = `
<nav
  className="f11f0r7k"
>
  <ul>
    <li>
      <Connect(Link)
        name="home"
      >
        <FormattedMessage
          defaultMessage="Home"
          id="home"
          values={Object {}}
        />
      </Connect(Link)>
    </li>
    <li>
      <Connect(Link)
        name="about"
      >
        <FormattedMessage
          defaultMessage="About Us"
          id="about"
          values={Object {}}
        />
      </Connect(Link)>
    </li>
    <li>
      <Connect(Link)
        name="counter"
      >
        <FormattedMessage
          defaultMessage="Counter"
          id="counter"
          values={Object {}}
        />
      </Connect(Link)>
    </li>
    <li>
      <Connect(Link)
        name="stars"
      >
        <FormattedMessage
          defaultMessage="Stars"
          id="stars"
          values={Object {}}
        />
      </Connect(Link)>
    </li>
    <li>
      <Connect(Link)
        name="register"
      >
        <FormattedMessage
          defaultMessage="Register"
          id="register"
          values={Object {}}
        />
      </Connect(Link)>
    </li>
  </ul>
</nav>
`;
